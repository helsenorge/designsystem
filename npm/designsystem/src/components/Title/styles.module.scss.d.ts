export type Styles = {
  title: string;
  'title--feature': string;
  'title--title1': string;
  'title--title2': string;
  'title--title3': string;
  'title--title4': string;
  'title--title5': string;
  'title--title6': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
