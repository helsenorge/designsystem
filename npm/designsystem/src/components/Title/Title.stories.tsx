import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import Title from './Title';
import { allTitleTags, allTitleAppearances } from '../../../.storybook/knobs';

const meta = {
  title: '@helsenorge∕designsystem-react/Components/Title',
  component: Title,
  parameters: {
    docs: {
      description: {
        component: 'Som innbygger vil jeg kunne vise titler i ulike størrelser og html markup.',
      },
    },
  },
  args: {
    children: 'Title',
    htmlMarkup: 'h1',
    margin: 0,
    appearance: 'title1',
  },
  argTypes: {
    children: {
      control: 'text',
    },
    htmlMarkup: {
      control: 'select',
      options: allTitleTags,
    },
    margin: {
      control: 'number',
    },
    appearance: {
      control: 'select',
      options: allTitleAppearances,
    },
  },
} satisfies Meta<typeof Title>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => <Title {...args} />,
};

export const AllAppearances: Story = {
  render: args => (
    <>
      <Title {...args} appearance="titleFeature">
        {args.children} {'(feature)'}
      </Title>
      <Title {...args} appearance="title1">
        {args.children} {'(title1)'}
      </Title>
      <Title {...args} appearance="title2">
        {args.children} {'(title2)'}
      </Title>
      <Title {...args} appearance="title3">
        {args.children} {'(title3)'}
      </Title>
      <Title {...args} appearance="title4">
        {args.children} {'(title4)'}
      </Title>
      <Title {...args} appearance="title5">
        {args.children} {'(title5)'}
      </Title>
    </>
  ),
};
