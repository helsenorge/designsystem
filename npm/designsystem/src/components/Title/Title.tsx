import React from 'react';

import classNames from 'classnames';

import { AnalyticsId } from '../../constants';

import titleStyles from './styles.module.scss';

export type TitleTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span';
export type TitleAppearances = 'titleFeature' | 'title1' | 'title2' | 'title3' | 'title4' | 'title5' | 'title6';

export interface TitleProps {
  children: React.ReactNode;
  /** Gives a unique id to the title */
  id?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Adds top and bottom margin in rem. */
  margin?: number | TitleMargin;
  /** Changes the underlying element of the title. */
  htmlMarkup?: TitleTags;
  /** Changes the appearance of the title. */
  appearance?: TitleAppearances;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export const Title = React.forwardRef(function TitleForwardedRef(props: TitleProps, ref: React.ForwardedRef<HTMLHeadingElement>) {
  const { id, children, className, htmlMarkup = 'h1', appearance = 'title1', margin = 0, testId } = props;
  const titleClasses = classNames(
    titleStyles.title,
    {
      [titleStyles['title--feature']]: appearance === 'titleFeature',
      [titleStyles['title--title1']]: appearance === 'title1',
      [titleStyles['title--title2']]: appearance === 'title2',
      [titleStyles['title--title3']]: appearance === 'title3',
      [titleStyles['title--title4']]: appearance === 'title4',
      [titleStyles['title--title5']]: appearance === 'title5',
      [titleStyles['title--title6']]: appearance === 'title6',
    },
    className
  );
  const CustomTag = htmlMarkup;

  const inlineStyle = instanceOfTitleMargin(margin)
    ? { marginTop: `${margin.marginTop}rem`, marginBottom: `${margin.marginBottom}rem` }
    : { marginTop: `${margin}rem`, marginBottom: `${margin}rem` };

  return (
    <CustomTag id={id} className={titleClasses} style={inlineStyle} ref={ref} data-testid={testId} data-analyticsid={AnalyticsId.Title}>
      {children}
    </CustomTag>
  );
});

export interface TitleMargin {
  marginTop: number;
  marginBottom: number;
}

export const instanceOfTitleMargin = (margin: unknown): margin is TitleMargin => {
  return Object.prototype.hasOwnProperty.call(margin, 'marginTop') && Object.prototype.hasOwnProperty.call(margin, 'marginBottom');
};

export default Title;
