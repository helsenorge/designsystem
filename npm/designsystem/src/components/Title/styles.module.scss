@import '../../scss/font-mixins';

.title {
  overflow-wrap: break-word;
  word-wrap: break-word;
  margin-left: 0;
  margin-right: 0;

  &--feature {
    @include title-feature;
  }

  &--title1 {
    @include title1;
  }

  &--title2 {
    @include title2;
  }

  &--title3 {
    @include title3;
  }

  &--title4 {
    @include title4;
  }

  &--title5 {
    @include title5;
  }

  &--title6 {
    @include title6;
  }
}
