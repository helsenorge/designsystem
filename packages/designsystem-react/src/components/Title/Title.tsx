import React from 'react';
import StyledTitle from './Title.styled';

export type TitleTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'span';
export type TitleAppearances = 'titleFeature' | 'title1' | 'title2' | 'title3' | 'title4' | 'title5';

interface TitleProps {
  children: React.ReactNode;
  id?: string;
  className?: string;
  margin?: number;
  htmlMarkup?: TitleTags;
  appearance?: TitleAppearances;
}

const Title = React.forwardRef((props: TitleProps, ref: any) => {
  const {id, children, className, htmlMarkup = 'h1', appearance = 'title1', margin = 0} = props;
  return (
    <StyledTitle id={id} className={className} as={htmlMarkup} margin={margin} appearance={appearance} ref={ref}>
      {children}
    </StyledTitle>
  );
});

export default Title;
