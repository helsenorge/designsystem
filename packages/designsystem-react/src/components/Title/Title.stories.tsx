import React from 'react';
import {withKnobs, text, select, number, boolean} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import Title from './Title';
import {withA11y} from '@storybook/addon-a11y';
import {allTitleTags, allTitleAppearances} from '../../../.storybook/knobs';

const stories = storiesOf('Title', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <div style={{width: '20rem'}}>
    <Title
      htmlMarkup={select('HtmlMarkup', allTitleTags, 'h1')}
      margin={number('Margin (rem)', 0)}
      appearance={select('Apperance', allTitleAppearances, 'title1')}>
      {text('Text', 'Title')}
    </Title>
  </div>
));

stories.add('All appearances', () => (
  <div>
    <Title
      htmlMarkup={select('HtmlMarkup', allTitleTags, 'h1')}
      margin={number('Margin (rem)', 0)}
      appearance="titleFeature">
      {`${text('Text', 'Title')} (feature)`}
    </Title>
    <Title htmlMarkup={select('HtmlMarkup', allTitleTags, 'h1')} margin={number('Margin (rem)', 0)} appearance="title1">
      {`${text('Text', 'Title')} (title1)`}
    </Title>
    <Title htmlMarkup={select('HtmlMarkup', allTitleTags, 'h1')} margin={number('Margin (rem)', 0)} appearance="title2">
      {`${text('Text', 'Title')} (title2)`}
    </Title>
    <Title htmlMarkup={select('HtmlMarkup', allTitleTags, 'h1')} margin={number('Margin (rem)', 0)} appearance="title3">
      {`${text('Text', 'Title')} (title3)`}
    </Title>
    <Title htmlMarkup={select('HtmlMarkup', allTitleTags, 'h1')} margin={number('Margin (rem)', 0)} appearance="title4">
      {`${text('Text', 'Title')} (title4)`}
    </Title>
    <Title htmlMarkup={select('HtmlMarkup', allTitleTags, 'h1')} margin={number('Margin (rem)', 0)} appearance="title5">
      {`${text('Text', 'Title')} (title5)`}
    </Title>
  </div>
));
