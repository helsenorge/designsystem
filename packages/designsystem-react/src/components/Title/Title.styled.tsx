import styled from 'styled-components';
import {TitleAppearances} from './Title';
import {typography} from '../../theme/typography';

export interface StyledTitleProps {
  appearance: TitleAppearances;
  margin: number;
}

const StyledTitle = styled('h1')<StyledTitleProps>`
  overflow-wrap: break-word;
  word-wrap: break-word;
  text-align: left;
  margin: ${props => `${props.margin}rem`} 0;
  ${props => typography[props.appearance]};
`;

export default StyledTitle;
