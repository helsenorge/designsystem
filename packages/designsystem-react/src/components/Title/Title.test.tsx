import React from 'react';
import {render} from '@testing-library/react';
import Title from './Title';

test('displays h1 default', (): void => {
  const {container} = render(<Title>Title</Title>);
  expect(container).toMatchSnapshot();
});
