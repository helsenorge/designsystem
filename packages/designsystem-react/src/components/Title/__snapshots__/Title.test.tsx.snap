// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays h1 default 1`] = `
<div>
  <h1
    class="sc-AxjAm fpVOuj"
  >
    Title
  </h1>
</div>
`;
